In [1]:
import pandas as pd
import seaborn as sns
%matplotlib inline
df = sns.load_dataset("tips")
df.shape

(244, 7)

In [0]:
df.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

## 1. total_bill 을 size 로 나눠서 1인당 가격을 구한결과를 "per_bill"이라는 컬럼에 담아주세요. 그리고 per_bill 컬럼에 담긴 값을 describe() 합니다.
```
count    244.00000
mean       7.88823
std        2.91435
min        2.87500
25%        5.80250
50%        7.25500
75%        9.39000
max       20.27500
Name: per_bill, dtype: float64
```

In [5]:
df["per_bill"] = df["total_bill"]/df["size"]
df["per_bill"].describe()

count    244.00000
mean       7.88823
std        2.91435
min        2.87500
25%        5.80250
50%        7.25500
75%        9.39000
max       20.27500
Name: per_bill, dtype: float64

## 2. groupby 를 사용해서 "time", "smoker" 를 그룹화 하고 "per_bill" 의 요약값을 구합니다. 
* 다음의 결과가 출력되도록 합니다.
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th>count</th>
      <th>mean</th>
      <th>std</th>
      <th>min</th>
      <th>25%</th>
      <th>50%</th>
      <th>75%</th>
      <th>max</th>
    </tr>
    <tr>
      <th>time</th>
      <th>smoker</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="2" valign="top">Lunch</th>
      <th>Yes</th>
      <td>23.0</td>
      <td>7.939420</td>
      <td>2.805047</td>
      <td>5.045000</td>
      <td>6.303333</td>
      <td>6.755000</td>
      <td>8.4350</td>
      <td>16.340</td>
    </tr>
    <tr>
      <th>No</th>
      <td>45.0</td>
      <td>6.996419</td>
      <td>2.082856</td>
      <td>3.755000</td>
      <td>5.585000</td>
      <td>6.580000</td>
      <td>8.3300</td>
      <td>12.355</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">Dinner</th>
      <th>Yes</th>
      <td>70.0</td>
      <td>9.003529</td>
      <td>3.839788</td>
      <td>2.875000</td>
      <td>6.060500</td>
      <td>7.875000</td>
      <td>11.1725</td>
      <td>20.275</td>
    </tr>
    <tr>
      <th>No</th>
      <td>106.0</td>
      <td>7.519203</td>
      <td>2.278733</td>
      <td>3.443333</td>
      <td>5.705833</td>
      <td>7.019167</td>
      <td>9.0050</td>
      <td>13.205</td>
    </tr>
  </tbody>
</table>

In [9]:
# 결과를 group_per 변수에 담고 group_per 변수를 출력해 주세요.
df.groupby(["time","smoker"])["per_bill"].describe()

count      mean       std       min       25%       50%  \
time   smoker                                                            
Lunch  Yes      23.0  7.939420  2.805047  5.045000  6.303333  6.755000   
       No       45.0  6.996419  2.082856  3.755000  5.585000  6.580000   
Dinner Yes      70.0  9.003529  3.839788  2.875000  6.060500  7.875000   
       No      106.0  7.519203  2.278733  3.443333  5.705833  7.019167   

                   75%     max  
time   smoker                   
Lunch  Yes      8.4350  16.340  
       No       8.3300  12.355  
Dinner Yes     11.1725  20.275  
       No       9.0050  13.205

## 3. 위 문제의 결과로 group_per 라고 저장된 데이터 프레임에서  Lunch 데이터만 가져옵니다.
* 아래의 결과가 출력되도록 합니다.
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>count</th>
      <th>mean</th>
      <th>std</th>
      <th>min</th>
      <th>25%</th>
      <th>50%</th>
      <th>75%</th>
      <th>max</th>
    </tr>
    <tr>
      <th>smoker</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Yes</th>
      <td>23.0</td>
      <td>7.939420</td>
      <td>2.805047</td>
      <td>5.045</td>
      <td>6.303333</td>
      <td>6.755</td>
      <td>8.435</td>
      <td>16.340</td>
    </tr>
    <tr>
      <th>No</th>
      <td>45.0</td>
      <td>6.996419</td>
      <td>2.082856</td>
      <td>3.755</td>
      <td>5.585000</td>
      <td>6.580</td>
      <td>8.330</td>
      <td>12.355</td>
    </tr>
  </tbody>
</table>

In [28]:
group_per = df.groupby(["time","smoker"])["per_bill"].describe()

In [29]:
group_per

count      mean       std       min       25%       50%  \
time   smoker                                                            
Lunch  Yes      23.0  7.939420  2.805047  5.045000  6.303333  6.755000   
       No       45.0  6.996419  2.082856  3.755000  5.585000  6.580000   
Dinner Yes      70.0  9.003529  3.839788  2.875000  6.060500  7.875000   
       No      106.0  7.519203  2.278733  3.443333  5.705833  7.019167   

                   75%     max  
time   smoker                   
Lunch  Yes      8.4350  16.340  
       No       8.3300  12.355  
Dinner Yes     11.1725  20.275  
       No       9.0050  13.205

## 4. "day", "smoker" 로 그룹화 하고 "size" 의 "count", "mean" 값을 구해주세요. 
* 다음의 결과가 출력되도록 합니다.
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th>count</th>
      <th>mean</th>
    </tr>
    <tr>
      <th>day</th>
      <th>smoker</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="2" valign="top">Thur</th>
      <th>Yes</th>
      <td>17</td>
      <td>2.352941</td>
    </tr>
    <tr>
      <th>No</th>
      <td>45</td>
      <td>2.488889</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">Fri</th>
      <th>Yes</th>
      <td>15</td>
      <td>2.066667</td>
    </tr>
    <tr>
      <th>No</th>
      <td>4</td>
      <td>2.250000</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">Sat</th>
      <th>Yes</th>
      <td>42</td>
      <td>2.476190</td>
    </tr>
    <tr>
      <th>No</th>
      <td>45</td>
      <td>2.555556</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">Sun</th>
      <th>Yes</th>
      <td>19</td>
      <td>2.578947</td>
    </tr>
    <tr>
      <th>No</th>
      <td>57</td>
      <td>2.929825</td>
    </tr>
  </tbody>
</table>

In [57]:
df.groupby(["day","smoker"])["size"].agg(["count", "mean"])

count      mean
day  smoker                 
Thur Yes        17  2.352941
     No         45  2.488889
Fri  Yes        15  2.066667
     No          4  2.250000
Sat  Yes        42  2.476190
     No         45  2.555556
Sun  Yes        19  2.578947
     No         57  2.929825

In [58]:
# 데이터를 로드해 주세요.
mpg = sns.load_dataset("mpg")
print(mpg.shape)
mpg.head()

(398, 9)


mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino

## 5. horsepower 가 결측치가 아닌 것만 가져와서 origin 으로 그룹화 하고 weight 의 평균값을 구합니다.
* 다음의 결과가 출력되도록 합니다.
```
origin
europe    2433.470588
japan     2221.227848
usa       3372.489796
Name: weight, dtype: float64
```

In [83]:
mpg.dropna().groupby("origin")["weight"].mean()

origin
europe    2433.470588
japan     2221.227848
usa       3372.489796
Name: weight, dtype: float64

## 6. name 에 audi 가 들어가는 데이터의 horsepower 컬럼의 요약값을 구합니다.
* 다음의 결과가 출력되도록 합니다.
```
count      7.000000
mean      86.714286
std       11.842217
min       67.000000
25%       80.500000
50%       90.000000
75%       93.000000
max      103.000000
Name: horsepower, dtype: float64
```

In [ ]:
df_seoul.loc[df_seoul["상호명_소문자"].str.contains("이디야|이디아|ediya"), "상호명_소문자"].shape

In [95]:
mpg[mpg["name"].str.contains("audi", "name")]["horsepower"].describe()

count      7.000000
mean      86.714286
std       11.842217
min       67.000000
25%       80.500000
50%       90.000000
75%       93.000000
max      103.000000
Name: horsepower, dtype: float64

## 7. name 에 diesel 이 들어가는 데이터에서 "origin", "model_year" 로 그룹화 하고 "horsepower" 의 평균을 구합니다.
* 다음의 결과가 출력되도록 합니다.

```
origin  model_year
europe  78            48.000000
        80            54.333333
        81            78.000000
usa     82            85.000000
Name: horsepower, dtype: float64
```

In [101]:
mpg[mpg["name"].str.contains("diesel","name")].groupby(["origin","model_year"])["horsepower"].mean()

origin  model_year
europe  78            48.000000
        80            54.333333
        81            78.000000
usa     82            85.000000
Name: horsepower, dtype: float64

## 8. "model_year"가 75보다 작은 값을 가지는 데이터프레임에서 pivot_table 을 사용해서 "model_year", "origin" 별  "weight"의 평균을 구해 다음과 같이 출력되도록 합니다.
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>origin</th>
      <th>europe</th>
      <th>japan</th>
      <th>usa</th>
    </tr>
    <tr>
      <th>model_year</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>70</th>
      <td>2309.200000</td>
      <td>2251.00</td>
      <td>3716.500000</td>
    </tr>
    <tr>
      <th>71</th>
      <td>2024.000000</td>
      <td>1936.00</td>
      <td>3401.600000</td>
    </tr>
    <tr>
      <th>72</th>
      <td>2573.200000</td>
      <td>2300.40</td>
      <td>3682.666667</td>
    </tr>
    <tr>
      <th>73</th>
      <td>2335.714286</td>
      <td>2397.25</td>
      <td>3821.448276</td>
    </tr>
    <tr>
      <th>74</th>
      <td>2139.333333</td>
      <td>2053.00</td>
      <td>3503.333333</td>
    </tr>
  </tbody>
</table>

In [126]:
pd.pivot_table(data = data,
               columns= "origin",
              index = "model_year",
              values = "weight", aggfunc = "mean")

origin           europe    japan          usa
model_year                                   
70          2309.200000  2251.00  3716.500000
71          2024.000000  1936.00  3401.600000
72          2573.200000  2300.40  3682.666667
73          2335.714286  2397.25  3821.448276
74          2139.333333  2053.00  3503.333333

## 9. x축에 horsepower 를 y축에 weight 를 지정하고 cylinders 에 따라 색상이 다르게 표현되도록 그래프를 그려주세요.
<img src="https://i.imgur.com/bMmZ2oi.png" width="400">

In [ ]:
ax = df.plot.scatter(x="a", y="b", color="DarkBlue", label="Group 1")
df.plot.scatter(x="c", y="d", color="DarkGreen", label="Group 2", ax=ax)

## 10. 위에서 불러온 mpg 데이터를 csv 파일로 저장합니다. (제출 시 저장한 csv 파일은 제출하지 않아도 됩니다.)
또, 인덱스 컬럼은 저장되지 않도록 합니다.

In [149]:
mpg.to_csv("mpg.csv")

In [153]:
# 제대로 저장이 되었다면 다음 형태로 파일이 불러와 집니다.
df = pd.read_csv("mpg.csv")
df

Unnamed: 0   mpg  cylinders  displacement  horsepower  weight  \
0             0  18.0          8         307.0       130.0    3504   
1             1  15.0          8         350.0       165.0    3693   
2             2  18.0          8         318.0       150.0    3436   
3             3  16.0          8         304.0       150.0    3433   
4             4  17.0          8         302.0       140.0    3449   
..          ...   ...        ...           ...         ...     ...   
393         393  27.0          4         140.0        86.0    2790   
394         394  44.0          4          97.0        52.0    2130   
395         395  32.0          4         135.0        84.0    2295   
396         396  28.0          4         120.0        79.0    2625   
397         397  31.0          4         119.0        82.0    2720   

     acceleration  model_year  origin                       name  
0            12.0          70     usa  chevrolet chevelle malibu  
1            11.5          70     usa          buick skylark 320  
2            11.0          70     usa         plymouth satellite  
3            12.0          70     usa              amc rebel sst  
4            10.5          70     usa                ford torino  
..            ...         ...     ...                        ...  
393          15.6          82     usa            ford mustang gl  
394          24.6          82  europe                  vw pickup  
395          11.6          82     usa              dodge rampage  
396          18.6          82     usa                ford ranger  
397          19.4          82     usa                 chevy s-10  

[398 rows x 10 columns]